lib/streamlit/runtime/caching/cached_message_replay.py

```diff
@@ -260,19 +254,21 @@ def calling_cached_function(
         It allows us to track any `st.foo` messages that are generated from inside the function
         for playback during cache retrieval.
         """
-        self._cached_func_stack.append(func)
         self._cached_message_stack.append([])
         self._seen_dg_stack.append(set())
-        if allow_widgets:
-            self._allow_widgets += 1
```

Comment:
Do you have context on why this used to be an int counter? Was there a good reason for it or is the change to boolean safe? I am wondering why it has not been a boolean before

Reply:
I reached out about that to @AnOctopus on Slack. My assumption is that in the current version, we only want to allow a single widget within a cached function since more widgets might create more variety to explode the cache. 
However, since the usage of `experimental_allow_widgets` is very low (only 0.5% of `cache_data` uses have this on `True), I'm not sure if it is worth the effort to implement this additional check with the new logic. The main aspect we need to prevent is usage of widgets inside `cache_data` if `experimental_allow_widgets==False`. This check feels a bit more like a nice to have, and also feels a bit unexpected since we are not documenting that anywhere that `experimental_allow_widgets` only allows a single widget.

Reply:
I got a response. This is intended for handling nested cached functions. I updated this PR + added a test to also allow nesting cached function. Thereby, if the outer function disallows widget replay it will also be disallowed for all inner functions.